libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp

```diff
@@ -0,0 +1,469 @@
+//  Copyright (c) 2021 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/allocator_support/internal_allocator.hpp>
+#include <hpx/execution_base/receiver.hpp>
+#include <hpx/functional/unique_function.hpp>
+#include <hpx/synchronization/mutex.hpp>
+
+#include <boost/container/small_vector.hpp>
+
+#include <memory>
+#include <mutex>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace experimental {
+    namespace detail {
+        enum class async_rw_mutex_access_type
+        {
+            read,
+            readwrite
+        };
+
+        struct async_rw_mutex_shared_state
+        {
+            hpx::lcos::local::mutex mtx;
+            boost::container::small_vector<
+                hpx::util::unique_function_nonser<void()>, 1>
+                continuations;
+
+            async_rw_mutex_shared_state() = default;
+            async_rw_mutex_shared_state(async_rw_mutex_shared_state&&) = delete;
+            async_rw_mutex_shared_state& operator=(
+                async_rw_mutex_shared_state&&) = delete;
+            async_rw_mutex_shared_state(
+                async_rw_mutex_shared_state const&) = delete;
+            async_rw_mutex_shared_state& operator=(
+                async_rw_mutex_shared_state const&) = delete;
+
+            ~async_rw_mutex_shared_state()
+            {
+                for (auto& continuation : continuations)
+                {
+                    HPX_INVOKE(std::move(continuation));
+                }
+            }
+
+            template <typename F>
+            void add_continuation(F&& continuation)
+            {
+                std::lock_guard<hpx::lcos::local::mutex> l(mtx);
+                continuations.emplace_back(std::move(continuation));
```

Comment:
Shouldn't this be:
```suggestion
                continuations.emplace_back(std::forward<F>(continuation));
```
?

Reply:
Sort of. I wanted to constrain the signature to only take rvalue references, but this also works.